### PR TITLE
make compilable with mingw64

### PR DESCRIPTION
**Pull request by eudoxos**
* see #416 for initial discussion
* removed by meson: src/arv{,gv}fakecamera.{c,h}, tests/realtimetest.c
* all optional features must be disabled (gst, usb, ...)
* all compilation errors are in arvgvinterface.c only (no ifaddrs.h)
* linux build not broken

### CODE REVIEW

**Comment by EmmanuelP** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/arvgvinterface.c  1
- src/arvmisc.c  1
         

Clones removed
==============
+ tests/arvcameratest.c  -1
+ tests/arvroitest.c  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/EmmanuelP/aravis/pullRequest?prid=6312336&bid=20524241)

**Comment by eudoxos** [comment]
Please have a look.

I replaced all enumerations with arv_enumerate_network_interfaces (the `_ARV_IFACES` is there to keep the old code around, for the moment). I tested in a limited way: arv-tool works locally.

There are few other changes related to a different plaform: for printing g_uint64, I use g_print instead of printf to avoid warnings.

I made separate changesets for various changes so that it is easier to review them.

File: src/arvgvfakecamera.c [hunk 1]
```diff
@@ -493,14 +501,27 @@ arv_gv_fake_camera_start (ArvGvFakeCamera *gv_fake_camera)
...
 			GInetAddress *inet_address;
 			GInetAddress *gvcp_inet_address;
 			unsigned int n_socket_fds;
 			unsigned int i;
-
+#ifndef _ARV_IFACES
 			socket_address = g_socket_address_new_from_native (ifap_iter->ifa_addr,
 									   sizeof (struct sockaddr));
+#else
+			socket_address = g_socket_address_new_from_native ((struct sockaddr*)((ArvIfaceAddr*)iface_iter->data)->addr,
```
Comment (EmmanuelP) [inline src/arvgvfakecamera.c:522]: Looks like struct sockaddr * cast is redundant. I would also prefer a getter function for addr.

File: src/arvgvfakecamera.c [hunk 2]
```diff
@@ -552,8 +577,11 @@ arv_gv_fake_camera_start (ArvGvFakeCamera *gv_fake_camera)
 			interface_found = TRUE;
 		}
 	}
-
+#ifndef _ARV_IFACES
 	freeifaddrs (ifap);
+#else
+	arv_enumerate_network_interfaces_free(ifaces);
```
Comment (EmmanuelP) [inline src/arvgvfakecamera.c:583]: enumerate returns a GList, please use g_lists_free_full here.

File: src/arvgvinterface.c [hunk 3]
```diff
@@ -25,24 +25,30 @@
...
 #include <arvmisc.h>
+#include <arvmiscprivate.h>
 #include <arvstr.h>
 #include <glib/gprintf.h>
 #include <gio/gio.h>
 #include <sys/types.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ifndef __MINGW32__
```
Comment (EmmanuelP) [inline src/arvgvinterface.c:42]: Let's try to put this in arvnetworkprivate.h

File: src/arvgvinterface.c [hunk 4]
```diff
@@ -71,8 +77,14 @@ arv_gv_discover_socket_set_buffer_size (ArvGvDiscoverSocket *discover_socket, gi
 	int result;
 
 	socket_fd = g_socket_get_fd (discover_socket->socket);
-
+#ifndef G_OS_WIN32
```
Comment (EmmanuelP) [inline src/arvgvinterface.c:80]: There is 2 occurences of this code, it is a call for an helper function in arvnetwork.c

File: src/arvgvstream.c [hunk 5]
```diff
@@ -34,10 +34,15 @@
 #include <arvgvcpprivate.h>
 #include <arvdebug.h>
 #include <arvmisc.h>
+#include <arvmiscprivate.h>
 #include <arvstr.h>
 #include <arvenumtypes.h>
 #include <string.h>
+#ifndef G_OS_WIN32
```
Comment (EmmanuelP) [inline src/arvgvstream.c:41]: -> arvnetworkprivate.h

File: src/arvmisc.c [hunk 6]
```diff
@@ -57,6 +62,79 @@ struct _ArvStatistic {
 	ArvHistogram *histograms;
 };
 
+#ifdef G_OS_WIN32
```
Comment (EmmanuelP) [inline src/arvmisc.c:65]: -> arvnetwork.c

File: src/arvmisc.c [hunk 7]
```diff
@@ -57,6 +62,79 @@ struct _ArvStatistic {
...
+GList* arv_enumerate_network_interfaces(void){
+	#warning NOT YET IMPLEMENTED
+	return NULL;
+}
+
+// http://mingw-users.1079350.n2.nabble.com/IPv6-getaddrinfo-amp-inet-ntop-td5891996.html
```
Comment (EmmanuelP) [inline src/arvmisc.c:71]: The linked page may not live forever, please add a bit of explanation on why this is needed.

File: src/arvmisc.c [hunk 8]
```diff
@@ -57,6 +62,79 @@ struct _ArvStatistic {
...
+	if(getifaddrs (&ifap) <0) return NULL;
+	for (ifap_iter = ifap; ifap_iter != NULL; ifap_iter = ifap_iter->ifa_next) {
+		if ((ifap_iter->ifa_flags & IFF_UP) != 0 &&
+			(ifap_iter->ifa_flags & IFF_POINTOPOINT) == 0 &&
+			(ifap_iter->ifa_addr != NULL) &&
+			(ifap_iter->ifa_addr->sa_family == AF_INET)) {
+				ArvIfaceAddr* a=(ArvIfaceAddr*)g_malloc(sizeof(ArvIfaceAddr));
+				memset(a,0,sizeof(ArvIfaceAddr));
+				a->addr=g_memdup(ifap_iter->ifa_addr,sizeof(struct sockaddr));
+				if(ifap_iter->ifa_netmask) a->netmask=g_memdup(ifap_iter->ifa_netmask,sizeof(struct sockaddr));
```
Comment (EmmanuelP) [inline src/arvmisc.c:110]: Coding style, code should be on a new line after the if statement.

File: src/arvmisc.c [hunk 9]
```diff
@@ -57,6 +62,79 @@ struct _ArvStatistic {
...
+		}
+	}
+	freeifaddrs (ifap);
+	ret=g_list_reverse(ret);
+	return ret;
+};
+#endif
+
+
+void _arv_iface_addr_free(ArvIfaceAddr*);
```
Comment (EmmanuelP) [inline src/arvmisc.c:123]: Please define a boxed type, like what is done for ArvValue.

File: src/arvmisc.c [hunk 10]
```diff
@@ -57,6 +62,79 @@ struct _ArvStatistic {
...
+
+void _arv_iface_addr_free(ArvIfaceAddr* a){
+	if(a->addr) g_free(a->addr);
```
Comment (EmmanuelP) [inline src/arvmisc.c:126]: Please use g_clear_pointer.

File: src/arvmisc.c [hunk 11]
```diff
@@ -57,6 +62,79 @@ struct _ArvStatistic {
...
+	if(a->netmask) g_free(a->netmask);
+	if(a->broadaddr) g_free(a->broadaddr);
+	if(a->name) g_free(a->name);
+}
+
+void arv_enumerate_network_interfaces_free(GList* ifaces){
```
Comment (EmmanuelP) [inline src/arvmisc.c:132]: I prefer to explicitely use g_list_free_full instead of this function.

File: src/arvmiscprivate.h [hunk 12]
```diff
@@ -33,4 +35,13 @@ struct _ArvValue {
 	} data;
 };
 
+typedef struct _ArvIfaceAddr ArvIfaceAddr;
+
+struct _ArvIfaceAddr{
+	struct sockaddr *addr, *netmask, *broadaddr;
```
Comment (EmmanuelP) [inline src/arvmiscprivate.h:41]: One line per member definition please.
Don't make this structure public, it should go to arvnetwork.c.

File: src/arvmiscprivate.h [hunk 13]
```diff
@@ -33,4 +35,13 @@ struct _ArvValue {
 	} data;
 };
 
+typedef struct _ArvIfaceAddr ArvIfaceAddr;
```
Comment (EmmanuelP) [inline src/arvmiscprivate.h:38]: Please rename ArvIfaceAddr to ArvNetworkInterface.

**Review by EmmanuelP — CHANGES_REQUESTED** [review]
Hi,

Thanks.

The patch set looks good, minor some coding style and consistency issues.

In the end, I'd prefer if the code related to network goes into its own module arvnetworkprivate.h and arvnetwork.c.

Also I prefer to rename ArvIfaceAddr to ArvNetworkInterface.

**Comment by EmmanuelP** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/arvgvinterface.c  1
- src/arvnetwork.c  1
         

Clones removed
==============
+ tests/arvcameratest.c  -1
+ tests/arvroitest.c  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/EmmanuelP/aravis/pullRequest?prid=6312336&bid=20524241)

File: src/arvgvinterface.c (hunk 4 above)
Comment (eudoxos) [reply src/arvgvinterface.c]: added as `arv_socket_set_recv_buffer_size`, declared in `arvnetoworkprivate.h`.

File: src/arvgvfakecamera.c [hunk 14]
```diff
@@ -493,14 +501,27 @@ arv_gv_fake_camera_start (ArvGvFakeCamera *gv_fake_camera)
 		    (ifap_iter->ifa_flags & IFF_POINTOPOINT) == 0 &&
 		    (ifap_iter->ifa_addr->sa_family == AF_INET) &&
 		    g_strcmp0 (ifap_iter->ifa_name, gv_fake_camera->priv->interface_name) == 0) {
+#else
+	ifaces=arv_enumerate_network_interfaces();
```
[the remark this answers is not in the document]
Comment (eudoxos) [reply src/arvgvfakecamera.c:503]: Will do this later, promise :) (hope clang-format will help me)

File: src/arvgvfakecamera.c [hunk 15]
```diff
@@ -474,14 +477,19 @@ _create_and_bind_input_socket (GSocket **socket_out, const char *socket_name, GI
 static gboolean
 arv_gv_fake_camera_start (ArvGvFakeCamera *gv_fake_camera)
 {
+#ifndef _ARV_IFACES
```
[the remark this answers is not in the document]
Comment (eudoxos) [reply src/arvgvfakecamera.c:478]: Correct, I will remove it when the other major things settle.

File: meson.build [hunk 16]
```diff
@@ -50,6 +51,11 @@ if usb_dep.found()
   aravis_dependencies += usb_dep
 endif
 
+if host_machine.system()=='windows'
```
[the remark this answers is not in the document]
Comment (eudoxos) [reply meson.build:53]: yes, AFAICT. I added a comment in there.

File: src/arvgvfakecamera.c (hunk 1 above)
Comment (eudoxos) [reply src/arvgvfakecamera.c]: For the getter function, you have something like `arv_network_interface_get_{addr,netmask, broadaddr,name}` for the data members in mind?

**Comment by eudoxos** [comment]
Thanks for the detailed comments.

File: src/arvgvfakecamera.c (hunk 1 above)
Comment (EmmanuelP) [reply src/arvgvfakecamera.c]: Yes.

**Comment by EmmanuelP** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/arvgvinterface.c  1
- src/arvnetwork.c  1
         

Clones removed
==============
+ tests/arvroitest.c  -1
+ tests/arvcameratest.c  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/EmmanuelP/aravis/pullRequest?prid=6312336&bid=20524241)

**Comment by EmmanuelP** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/arvnetwork.c  1
- src/arvgvinterface.c  1
         

Clones removed
==============
+ tests/arvcameratest.c  -1
+ tests/arvroitest.c  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/EmmanuelP/aravis/pullRequest?prid=6312336&bid=20524241)

File: src/arvnetwork.h [hunk 17]
```diff
@@ -0,0 +1,56 @@
+/* Aravis - Digital camera library
```
Comment (EmmanuelP) [inline src/arvnetwork.h:1]: Please put everything in arvnetworkprivate.h, the API here are internal to aravis and should not be exported. No need for a arvnetwork.h file.

File: src/arvnetworkprivate.h [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+    u_int16_t check;
+    u_int32_t saddr;
+    u_int32_t daddr;
+    /*The options start here. */
+  };
+#endif
+
+
+
+struct _ArvNetworkInterface{
```
Comment (EmmanuelP) [inline src/arvnetworkprivate.h:89]: This definition should be in arvnetwork.c

File: src/arvnetwork.h (hunk 17 above)
Comment (eudoxos) [reply src/arvnetwork.h:1]: Now I understand; when we have accessors, there is no need to expose the internals at all.

**Comment by eudoxos** [comment]
Style was hand-fixed, old code removed, tests pass. From my side, all the points are addessed. More comments welcome. I'd prefer this to be merged in this state despite `arv_enumerate_network_interfaces` still being a stub for win32. This can be done later, however, and will be a localized addition only.

**Comment by EmmanuelP** [comment]
Hi,

> I'd prefer this to be merged in this state despite arv_enumerate_network_interfaces still being a stub for win32. This can be done later, however, and will be a localized addition only.

I'm OK with that.

I'm not seeing your changes. Did you forget to push them ?

**Comment by eudoxos** [comment]
I did forget, sorry. Pushed just now.

**Comment by EmmanuelP** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ tests/arvcameratest.c  -1
+ tests/arvroitest.c  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/EmmanuelP/aravis/pullRequest?prid=6312336&bid=20524241)

**Comment by EmmanuelP** [comment]
Hi, I have incorporated your changes in master without the last commit implementing interface enumeration for win32. Please create a new merge request for this work.

Thanks,

    Emmanuel.

**Comment by EmmanuelP** [comment]
Please run tests/valgrind-memcheck for leak checks.

Your patch set introduced 2 leaks:

982c6fd3329f65f134081839ee1f74238d85b9b5
15af691ba318118beaaa8e9c89ab35cb03c7d1bb

Thanks.

Cheers.

**Comment by eudoxos** [comment]
Thanks for merging, and for spotting & fixing those.

**Pull request closed without merging**